Sources/Sentry/SentryProfiler.mm

```diff
@@ -243,146 +241,195 @@
-    [profiler addTransaction:transaction];
+    NSMutableDictionary<NSString *, id> *payload = [_gCurrentProfiler->_profileData mutableCopy];
 
-    [self captureEnvelopeIfFinished:profiler spanID:spanID];
-}
+    NSArray *samples = ((NSArray *)payload[@"profile"][@"samples"]).mutableCopy;
 
-+ (BOOL)isRunning
-{
-    std::lock_guard<std::mutex> l(_gProfilerLock);
-    return [_gCurrentProfiler isRunning];
+    // We need at least two samples to be able to draw a stack frame for any given function: one
+    // sample for the start of the frame and another for the end. Otherwise we would only have a
+    // stack frame with 0 duration, which wouldn't make sense.
+    if ([samples count] < 2) {
+        SENTRY_LOG_DEBUG(@"Not enough samples in profile");
+        return nil;
+    }
+
+    // slice the profile data to only include the samples/metrics within the transaction
+    const auto slicedSamples = [self slicedArray:samples transaction:transaction];
+    if (slicedSamples.count < 2) {
+        SENTRY_LOG_DEBUG(@"Not enough samples in profile during the transaction");
+        return nil;
+    }
+    payload[@"profile"][@"samples"] = slicedSamples;
+
+    // add the serialized info for the associated transaction
+    const auto firstSampleTimestamp
+        = (uint64_t)[slicedSamples.firstObject[@"elapsed_since_start_ns"] longLongValue];
+    const auto profileDuration = getDurationNs(firstSampleTimestamp, getAbsoluteTime());
+
+    const auto transactionInfo = [self serializeInfoForTransaction:transaction
+                                                   profileDuration:profileDuration];
+    if (!transactionInfo) {
+        SENTRY_LOG_WARN(@"Could not find any associated transaction for the profile.");
+        return nil;
+    }
+    payload[@"transactions"] = @[ transactionInfo ];
+
+    // add the gathered metrics
+    NSMutableDictionary<NSString *, id> *metrics = [_gCurrentProfiler->_metricProfiler serialize];
+
+    const auto slicedMetrics =
+        [NSMutableDictionary<NSString *, id> dictionaryWithCapacity:metrics.count];
+    void (^metricSlicer)(NSString *_Nonnull, NSDictionary<NSString *, id> *_Nonnull, BOOL *_Nonnull)
+        = ^(NSString *_Nonnull metricKey, NSDictionary<NSString *, id> *_Nonnull nextMetricsEntry,
+            BOOL *_Nonnull metricStop) {
+              NSArray<NSDictionary<NSString *, NSString *> *> *nextMetrics
+                  = nextMetricsEntry[@"values"];
+
+              const auto nextSlicedMetrics = [self slicedArray:nextMetrics transaction:transaction];
+              if (!nextSlicedMetrics) {
+                  return;
+              }
+
+              slicedMetrics[metricKey] =
+                  @{ @"unit" : nextMetricsEntry[@"unit"], @"values" : nextSlicedMetrics };
+          };
+    [metrics enumerateKeysAndObjectsUsingBlock:metricSlicer];
+
+#    if SENTRY_HAS_UIKIT
+    const auto slowFrames
+        = processFrameRenders(_gCurrentFramesTracker.currentFrames.slowFrameTimestamps,
```

Comment:
`m`: We should only call `_gCurrentFramesTracker.currentFrames` once and store it as a local variable to ensure consistent data. Otherwise, it could happen that the `currentFrames` change during the subsequent calls here.

Reply:
This would only store the pointer to the same object that would continue to be modified while we deref it. We'd have to perform a copy, but I don't think that's necessary given that we're removing any unneeded data anyways, bounded to the transaction. So we can freely add to this data structure without worry.

We won't end up in a scenario where we grab slow frames and then the data structure is wiped before we can grab frozen frames for the same profile, as the data structure wont be reset until after this method finishes. That sequence is synchronized; frames are accessed in `createProfilingEnvelopeItemForTransaction`, called in `-[SentryTracer captureTransactionInEnvelope]` on line 538 which syncs with a `lock_guard`: https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryTracer.m#L538

then later on line 552 `stop` is called which resets the frames, and also syncs with a `lock_guard`:

https://github.com/getsentry/sentry-cocoa/blob/56141bd6dfb7d4908a40df537f7f56beb42a83cd/Sources/Sentry/SentryTracer.m#L552

So we should never be able to have one instance of SentryTracer modify the frames structure while another is accessing it via `SentryProfiler`.

Reply:
I missed that. Thanks for the detailed explanation.